db/src/metadata.rs

```diff
@@ -0,0 +1,379 @@
+                }
+            },
+
+            entids::DB_ALTER_ATTRIBUTE => {
+                if entid != entids::DB_PART_DB {
+                    bail!(ErrorKind::BadSchemaAssertion(format!("Expected [:db.part/db :db.alter/attribute ...] but got [{:?} :db.alter/attribute ...]", entid)))
+                }
+
+                match *value {
+                    TypedValue::Ref(e) => {
+                        altered.insert(e);
+                        continue
+                    },
+                    _ => bail!(ErrorKind::BadSchemaAssertion(format!("Expected [:db.part/db :db.alter/attribute entid] but got [:db.part/db :db.alter/attribute {:?}]", value)))
+                }
+            },
+            _ => {},
+        }
+
+        let builder = builders.entry(entid).or_insert(AttributeBuilder::default());
+
+        // TODO: improve error messages throughout.
+        match attr {
+            entids::DB_DOC => {
+                match *value {
+                    TypedValue::String(_) => {},
+                    _ => bail!(ErrorKind::BadSchemaAssertion(format!("Expected [... :db/doc \"string value\"] but got [... :db/doc {:?}] for entid '{}' and attribute '{}'", value, entid, attr)))
+                }
+            },
+
+            entids::DB_VALUE_TYPE => {
+                match *value {
+                    TypedValue::Ref(entids::DB_TYPE_REF) => { builder.value_type(ValueType::Ref); },
+                    TypedValue::Ref(entids::DB_TYPE_BOOLEAN) => { builder.value_type(ValueType::Boolean); },
+                    TypedValue::Ref(entids::DB_TYPE_DOUBLE) => { builder.value_type(ValueType::Double); },
+                    TypedValue::Ref(entids::DB_TYPE_LONG) => { builder.value_type(ValueType::Long); },
+                    TypedValue::Ref(entids::DB_TYPE_STRING) => { builder.value_type(ValueType::String); },
+                    TypedValue::Ref(entids::DB_TYPE_KEYWORD) => { builder.value_type(ValueType::Keyword); },
+                    _ => bail!(ErrorKind::BadSchemaAssertion(format!("Expected [... :db/valueType :db.type/*] but got [... :db/valueType {:?}] for entid '{}' and attribute '{}'", value, entid, attr)))
+                }
+            },
+
+            entids::DB_CARDINALITY => {
+                match *value {
+                    TypedValue::Ref(entids::DB_CARDINALITY_MANY) => { builder.multival(true); },
+                    TypedValue::Ref(entids::DB_CARDINALITY_ONE) => { builder.multival(false); },
+                    _ => bail!(ErrorKind::BadSchemaAssertion(format!("Expected [... :db/cardinality :db.cardinality/many|:db.cardinality/one] but got [... :db/cardinality {:?}]", value)))
+                }
+            },
+
+            entids::DB_UNIQUE => {
+                match *value {
+                    // TODO: accept nil in some form.
+                    // TypedValue::Nil => {
```

Comment:
I wasn't precise: by "some form" I was thinking perhaps `:db.unique/nil` or `:db/notUnique`, for precisely the reasons you note about `nil`.  I was not aware you could retract to unset; my current patch doesn't allow that.  I'll consider allowing it once I work out some of the details.